### PR TITLE
Silk.NET.Vulkan.Extensions.ImGui

### PR DESCRIPTION
**Pull request by bmdub**
This is an attempt to integrate my port of imgui_impl_vulkan.cpp into Silk as "Silk.NET.Vulkan.Extensions.ImGui".

Some notes thus far:
- I'm a bit troubled by how many parameters are needed to construct this ImGuiController class.  Kinda ugly.  Thanks, Vulkan.
- The example app seems to have an issue where the imgui windows will disappear while you are shrinking the window.  Dunno what's going on there / haven't dug into it.
- I'll be out tomorrow but feel free to comment whenever you have time

Thanks!

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481) before we can accept your contribution.<br/><br/>:x: bmdub [sign now](https://cla.dotnetfoundation.org/dotnet/Silk.NET?pullRequest=481)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/Silk.NET?pullRequest=481) it.</sub>

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+			io.KeyMap[(int)ImGuiKey.Y] = (int)Key.Y;
+			io.KeyMap[(int)ImGuiKey.Z] = (int)Key.Z;
+		}
+
+		private unsafe void RenderImDrawData(in ImDrawDataPtr drawDataPtr, in CommandBuffer commandBuffer, in Framebuffer framebuffer, in Extent2D swapChainExtent)
+		{
+			int framebufferWidth = (int)(drawDataPtr.DisplaySize.X * drawDataPtr.FramebufferScale.X);
+			int framebufferHeight = (int)(drawDataPtr.DisplaySize.Y * drawDataPtr.FramebufferScale.Y);
+			if (framebufferWidth <= 0 || framebufferHeight <= 0)
+				return;
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs:401]: Fix branchless ifs in whole file. Generally, I prefer to have branches even for single-line ifs mainly to eliminate the potential for bugs.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 2]
```diff
@@ -0,0 +1,1453 @@
+//////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:1]: Doesn't start with the standard file header.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs [hunk 3]
```diff
@@ -0,0 +1,435 @@
+using System;
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/ImGuiController.cs:1]: Doesn't start with the standard file header (fix this in all files you've added, even the GLSL ones)

File: src/Lab/ImGuiVulkan/ImGuiVulkanApplication.cs [hunk 4]
```diff
@@ -0,0 +1,1211 @@
...
+using Silk.NET.Vulkan.Extensions.ImGui;
+using Silk.NET.Vulkan.Extensions.KHR;
+using Silk.NET.Windowing;
+using Image = Silk.NET.Vulkan.Image;
+
+namespace ImGuiVulkan
+{
+	public class ImGuiVulkanApplication
+	{
+		public const bool EnableValidationLayers = true;
```
Comment (Perksey) [inline src/Lab/ImGuiVulkan/ImGuiVulkanApplication.cs:23]: After a reformat this file will be fine, we're not as fussy over lab applications so long as they work.

File: src/Lab/ImGuiVulkan/Program.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+using System.IO;
+
+namespace ImGuiVulkan
+{
+	class Program
+	{
+		static void Main(string[] args)
+		{
+			new ImGuiVulkanApplication().Run();
+        }
```
Comment (Perksey) [inline src/Lab/ImGuiVulkan/Program.cs:11]: Formatting is super wonky here.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 6]
```diff
@@ -0,0 +1,1453 @@
...
+//
+// This is a C# port of the Vulkan imgui backend implementation:
+// https://github.com/ocornut/imgui/blob/master/backends/imgui_impl_vulkan.h
+// https://github.com/ocornut/imgui/blob/master/backends/imgui_impl_vulkan.cpp
+// (latest commit: https://github.com/ocornut/imgui/commit/6d3a980f38d849e9eac22180dce12f4c2e16794c)
+//
+// The names, layout, and overall convention was preserved as much as possible (almost), in order to make
+// future updates easy to apply.
+// Exceptions: Some unused methods were removed, as well as unused members of ImGui_ImplVulkan_InitInfo.
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:10]: Would rather align this with the rest of the codebase, with C#-style names and everything else.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 7]
```diff
@@ -0,0 +1,1453 @@
...
+//  2018-03-03: Vulkan: Various refactor, created a couple of ImGui_ImplVulkanH_XXX helper that the example can use and that viewport support will use.
+//  2018-03-01: Vulkan: Renamed ImGui_ImplVulkan_Init_Info to ImGui_ImplVulkan_InitInfo and fields to match more closely Vulkan terminology.
+//  2018-02-16: Misc: Obsoleted the io.RenderDrawListsFn callback, ImGui_ImplVulkan_Render() calls ImGui_ImplVulkan_RenderDrawData() itself.
+//  2018-02-06: Misc: Removed call to ImGui::Shutdown() which is not available from 1.60 WIP, user needs to call CreateContext/DestroyContext themselves.
+//  2017-05-15: Vulkan: Fix scissor offset being negative. Fix new Vulkan validation warnings. Set required depth member for buffer image copy.
+//  2016-11-13: Vulkan: Fix validation layer warnings and errors and redeclare gl_PerVertex.
+//  2016-10-18: Vulkan: Add location decorators & change to use structs as in/out in glsl, update embedded spv (produced with glslangValidator -x). Null the released resources.
+//  2016-08-27: Vulkan: Fix Vulkan example for use when a depth buffer is active.
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:75]: Is this really necessary

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 8]
```diff
@@ -0,0 +1,1453 @@
...
+		// Font data
+		static Sampler g_FontSampler = default;
+		static DeviceMemory g_FontMemory = default;
+		static Image g_FontImage = default;
+		static ImageView g_FontView = default;
+		static DeviceMemory g_UploadBufferMemory = default;
+		static Buffer g_UploadBuffer = default;
+
+		// Render buffers
+		static ImGui_ImplVulkanH_WindowRenderBuffers g_MainWindowRenderBuffers;
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:194]: We shouldn't be capturing static state.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 9]
```diff
@@ -0,0 +1,1453 @@
...
+		static DescriptorSetLayout g_DescriptorSetLayout = default;
+		static PipelineLayout g_PipelineLayout = default;
+		static DescriptorSet g_DescriptorSet = default;
+		static Pipeline g_Pipeline = default;
+		static uint g_Subpass = 0;
+		static ShaderModule g_ShaderModuleVert;
+		static ShaderModule g_ShaderModuleFrag;
+		static bool g_FunctionsLoaded = true;
+		static ImGui_ImplVulkanH_FrameRenderBuffers[] g_frameRenderBuffers;
+		static GCHandle g_frameRenderBuffersHandle;
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:179]: GCHandles are very very slow and we should remove them completely.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 10]
```diff
@@ -0,0 +1,1453 @@
...
+				vk.CmdBindPipeline(command_buffer, PipelineBindPoint.Graphics, pipeline);
+				Span<DescriptorSet> desc_set = stackalloc DescriptorSet[] { g_DescriptorSet };
+				vk.CmdBindDescriptorSets(command_buffer, PipelineBindPoint.Graphics, g_PipelineLayout, 0, 1, desc_set, 0, (uint*)null);
+			}
+
+			// Bind Vertex And Index Buffer:
+			if (draw_data.TotalVtxCount > 0)
+			{
+				ReadOnlySpan<Buffer> vertex_buffers = stackalloc Buffer[] { rb.VertexBuffer };
+				ulong[] vertex_offset = new ulong[] { 0 };
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:372]: Use stackalloc. In fact just use a local.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 11]
```diff
@@ -0,0 +1,1453 @@
...
+				vk.CmdBindVertexBuffers(command_buffer, 0, 1, vertex_buffers, vertex_offset);
+				vk.CmdBindIndexBuffer(command_buffer, rb.IndexBuffer, 0, sizeof(ushort) == 2 ? IndexType.Uint16 : IndexType.Uint32);
+			}
+
+			// Setup viewport:
+			{
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:378]: Why the bare scope?

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 12]
```diff
@@ -0,0 +1,1453 @@
...
+				range[0].Memory = g_UploadBufferMemory;
+				range[0].Size = upload_size;
+				err = vk.FlushMappedMemoryRanges(v.Device, 1, range);
+				check_vk_result(err);
+				vk.UnmapMemory(v.Device, g_UploadBufferMemory);
+			}
+
+			// Copy to Image:
+			{
+				Span<ImageMemoryBarrier> copy_barrier = stackalloc ImageMemoryBarrier[1];
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:648]: Why can't this just be a local, no stackalloc?

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 13]
```diff
@@ -0,0 +1,1453 @@
...
+			check_vk_result(err);
+		}
+
+		static unsafe void ImGui_ImplVulkan_CreateDescriptorSetLayout(Vk vk, Device device, AllocationCallbacks* allocator)
+		{
+			if (g_DescriptorSetLayout.Handle != default)
+				return;
+
+			ImGui_ImplVulkan_CreateFontSampler(vk, device, allocator);
+			fixed (Sampler* pSampler = stackalloc Sampler[1])
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:751]: This doesn't need to be fixed.

File: src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs [hunk 14]
```diff
@@ -0,0 +1,1453 @@
...
+			{
+				Span<DescriptorSetLayoutBinding> binding = stackalloc DescriptorSetLayoutBinding[1];
+				binding[0].DescriptorType = DescriptorType.CombinedImageSampler;
+				binding[0].DescriptorCount = 1;
+				binding[0].StageFlags = ShaderStageFlags.ShaderStageFragmentBit;
+				binding[0].PImmutableSamplers = pSampler;
+				fixed (DescriptorSetLayoutBinding* pBinding = binding)
```
Comment (Perksey) [inline src/Vulkan/Extensions/Silk.NET.Vulkan.Extensions.ImGui/imgui_impl_vulkan.cs:758]: This doesn't need to be fixed.

**Comment by bmdub** [comment]
can do!

**Comment by bmdub** [comment]
@Perksey here you go - 

- Moved library into Lab/ImGuiVulkan app for further iteration.
- I ended up refactoring it quite a bit, replacing the straight C -> C# conversion/abomination with something more proper.  It turns out some of that C code was doing duplicate operations!
- I cleaned up probably most of what you were commenting on.  Notably, I did not touch the exceptions because I wasn't confident on exactly how they should be.
- It seems Visual Studio has changed a bunch of guids in the .sln.  I can re-push without adding to the .sln if needed.  Also, I did not see the code spacing issues, but who knows what goes on with Visual Studio.

This should be a much better handing-off point for you guys :)  I think someone who knows Vulkan better than I can still improve on this.

**Review by Perksey — APPROVED** [review]
Awesome, thanks. Happy to approve for further iteration down the line.

**Pull request merged**